### ISSUE TITLE
TF Serving for "Distributed training and monitoring" by running d_traineval.ipynb gives {     "error": "Serving signature name: "serving_default" not found in signature def" }

### ISSUE THREAD

**Issue by amithbk12man** [GoogleCloudPlatform/training-data-analyst#380]
While i executed "d_traineval.ipynb" I could export the exported model. I used the model and used a local docker image of Tensorflow serving 1.8 CPU and i get the following result as output for REST post call
{
    "error": "Serving signature name: \"serving_default\" not found in signature def"
}

My request JSON:
{
  "instances": [
    {"pickuplon" : -73.987625,
    "pickuplat" : 40.750617,
    "dropofflat" : 40.78518,
    "dropofflon" : -73.971163,
    "passengers" : 2
    }
  ]
}

Can you please help me what is the error?

**Comment by amithbk12man** [comment]
Adding the below output of trained model.

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_save_checkpoints_secs': 600, '_num_ps_replicas': 0, '_keep_checkpoint_max': 5, '_task_type': 'worker', '_global_id_in_cluster': 0, '_is_chief': True, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x7fcc98460dd0>, '_model_dir': '/data/taxi_trained', '_protocol': None, '_save_checkpoints_steps': None, '_keep_checkpoint_every_n_hours': 10000, '_service': None, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_tf_random_seed': None, '_save_summary_steps': 100, '_device_fn': None, '_experimental_distribute': None, '_num_worker_replicas': 1, '_task_id': 0, '_log_step_count_steps': 100, '_evaluation_master': '', '_eval_distribute': None, '_train_distribute': None, '_master': ''}
INFO:tensorflow:Running training and evaluation locally (non-distributed).
INFO:tensorflow:Start train and evaluate loop. The evaluate will happen after every checkpoint. Checkpoint frequency is determined based on RunConfig arguments: save_checkpoints_steps None or save_checkpoints_secs 600.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 0 into /data/taxi_trained/model.ckpt.
INFO:tensorflow:loss = 94763.02, step = 1
INFO:tensorflow:global_step/sec: 34.996
INFO:tensorflow:loss = 45864.14, step = 101 (2.860 sec)
INFO:tensorflow:global_step/sec: 42.3265
INFO:tensorflow:loss = 42262.1, step = 201 (2.362 sec)
INFO:tensorflow:global_step/sec: 42.1929
INFO:tensorflow:loss = 53431.258, step = 301 (2.374 sec)
INFO:tensorflow:global_step/sec: 41.2542
INFO:tensorflow:loss = 43426.207, step = 401 (2.419 sec)
INFO:tensorflow:Saving checkpoints for 500 into /data/taxi_trained/model.ckpt.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2019-02-11-18:00:53
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from /data/taxi_trained/model.ckpt-500
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2019-02-11-18:00:54
INFO:tensorflow:Saving dict for global step 500: average_loss = 109.248985, global_step = 500, label/mean = 11.666427, loss = 45474.89, prediction/mean = 11.013848
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 500: /data/taxi_trained/model.ckpt-500
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Signatures INCLUDED in export for Eval: None
INFO:tensorflow:Signatures INCLUDED in export for Classify: None
INFO:tensorflow:Signatures INCLUDED in export for Regress: None
INFO:tensorflow:Signatures INCLUDED in export for Predict: ['predict']
INFO:tensorflow:Signatures INCLUDED in export for Train: None
INFO:tensorflow:Signatures EXCLUDED from export because they cannot be be served via TensorFlow Serving APIs:
INFO:tensorflow:'serving_default' : Regression input must be a single string Tensor; got {'passengers': <tf.Tensor 'Placeholder_4:0' shape=(?,) dtype=float32>, 'pickuplon': <tf.Tensor 'Placeholder:0' shape=(?,) dtype=float32>, 'dropofflon': <tf.Tensor 'Placeholder_3:0' shape=(?,) dtype=float32>, 'pickuplat': <tf.Tensor 'Placeholder_1:0' shape=(?,) dtype=float32>, 'dropofflat': <tf.Tensor 'Placeholder_2:0' shape=(?,) dtype=float32>}
INFO:tensorflow:'regression' : Regression input must be a single string Tensor; got {'passengers': <tf.Tensor 'Placeholder_4:0' shape=(?,) dtype=float32>, 'pickuplon': <tf.Tensor 'Placeholder:0' shape=(?,) dtype=float32>, 'dropofflon': <tf.Tensor 'Placeholder_3:0' shape=(?,) dtype=float32>, 'pickuplat': <tf.Tensor 'Placeholder_1:0' shape=(?,) dtype=float32>, 'dropofflat': <tf.Tensor 'Placeholder_2:0' shape=(?,) dtype=float32>}
WARNING:tensorflow:Export includes no default signature!
INFO:tensorflow:Restoring parameters from /data/taxi_trained/model.ckpt-500
WARNING:tensorflow:From /home/jupyter/libraries/nb_env/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py:1018: calling add_meta_graph_and_variables (from tensorflow.python.saved_model.builder_impl) with legacy_init_op is deprecated and will be removed in a future version.
Instructions for updating:
Pass your op to the equivalent parameter main_op instead.
INFO:tensorflow:Assets added to graph.
INFO:tensorflow:No assets to write.
INFO:tensorflow:SavedModel written to: /data/taxi_trained/export/exporter/temp-1549908054/saved_model.pb
INFO:tensorflow:Loss for final step: 39433.676.

**Comment by lakshmanok** [comment]
Run "saved_model_cli" on the exported model, and find out the actual signature (it might be called "predict" for example). then specify the signature_name attribute in the request:

https://www.tensorflow.org/serving/api_rest#request_format

**Comment by amithbk12man** [comment]
Hi Lak,
    Thanks for the prompt answer. I will run "saved_model_cli" and check. 
Just to add i use "predict" only and my rest API: <url>:predict and then i get 
{
    "error": "Serving signature name: \"predict\" not found in signature def"
}

**Comment by StephenZMZ** [comment]
Hi,all,I have the same problem!!! when I use tensorflow-serving prediction, return---->{ "error": "Serving signature name: \"serving_default\" not found in signature def" }
